daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala

```diff
@@ -463,6 +476,73 @@ class IdeClient(val compiledPackages: CompiledPackages) extends ScriptLedgerClie
     Future.fromTry(result)
   }
 
+  override def submitMustFail(party: SParty, commands: List[ScriptLedgerClient.Command])(
```

Comment:
I am slightly worried that we are effectively replicating a significant part of the logic of `submit`. What happens if we replace the whole implementation here by something that catches `SError` and flips it?

Reply:
By catching `SError` you mean the `SResultError` branch? AFAIU that's not enough as we also want to flip in other cases, e.g. on `commitTransaction` error in `SResultScenarioCommit`. Conversely we don't always want to recover from `Failure`, e.g. AFAIU we don't want to recover from `Failure(new RuntimeException(s"Unexpected abort: $x"))`. The other `submit` implementations use `Left` for errors that should be caught by `submitMustFail` but the `Left` parameter is restricted to `StatusRuntimeException` which doesn't really work in this case.

An alternative might be to use a more general or a trait associated type instead of `StatusRuntimeException` and then use `Success(Left(_))` in `IdeClient.submit` for errors that `submitMustFail` should catch.

Reply:
Side note: SResultScenarioCommit was a bug, fixed in https://github.com/digital-asset/daml/pull/7089. Doesn’t change anything here.

There are a few cases to consider here:
1. The future completes successfully. In that case, we want to turn it into a failure.
2. The future fails with `SError`. This includes calls to `error` as well as `ScenarioErrorCommitError` which we will get back from a failed commit for things like authorization errors.
3. Other stuff, e.g., the wildcard match at the end which we hope not to hit.

My idea would be to call `submit` and then handle the three cases.
1. This is an actual failure. We expected things to fail and they didn’t.
2. This is the successful case. We expected a failure during submission and we got one.
3. This is a crash, something went wrong.

Reply:
Ah, I see. That seems to be possible. I've pushed a commit implementing this.